File: shoring/flywheel/src/test/java/com/panopset/tests/flywheel/EscapeTest.java
package com.panopset.tests.flywheel;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import com.panopset.compat.Stringop;
import com.panopset.flywheel.Flywheel;
import com.panopset.flywheel.FlywheelBuilder;
import com.panopset.flywheel.LineFeedRules;

public class EscapeTest {

  @Test
  void testEscapeNewLine() throws IOException {
    Assertions.assertEquals("foobar\n",
        applyTemplate(Arrays.asList(new String[] {"foo\\", "bar"})));
    Assertions.assertEquals("\n", applyTemplate(Arrays.asList(new String[] {"\\", ""})));
    Assertions.assertEquals("\\\n\n", applyTemplate(Arrays.asList(new String[] {"\\\\", ""})));
    Assertions.assertEquals("\\\n", applyTemplate(Arrays.asList(new String[] {"\\\\"})));
    Assertions.assertEquals("foobar\n", applyTemplate(Arrays.asList(new String[] {"${@p x}foo\\", "bar${@q}${x}"})));
  }

  private String applyTemplate(List<String> input) {
    Stringop.setEol("\n");
    Flywheel fw =
        new FlywheelBuilder().input(input).withLineFeedRules(LineFeedRules.LINE_BREAKS).construct();
    String result = fw.exec();
    return result;
  }
}

File: code/shoring/flywheel/src/test/kotlin/com/panopset/tests/flywheel/EscapeTest.kt
package com.panopset.tests.flywheel

import com.panopset.compat.Stringop.setEol
import com.panopset.flywheel.FlywheelBuilder
import com.panopset.flywheel.LINE_BREAKS
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.*

class EscapeTest {
    @Test
    fun testEscapeNewLine() {
        Assertions.assertEquals(
            "foobar",
            applyTemplate(Arrays.asList(*arrayOf("foo\\", "bar")))
        )
        Assertions.assertEquals("", applyTemplate(Arrays.asList(*arrayOf("\\", ""))))
        Assertions.assertEquals("\\\n", applyTemplate(Arrays.asList(*arrayOf("\\\\", ""))))
        Assertions.assertEquals("\\", applyTemplate(Arrays.asList(*arrayOf("\\\\"))))
        Assertions.assertEquals("foobar", applyTemplate(Arrays.asList(*arrayOf("\${@p x}foo\\", "bar\${@q}\${x}"))))
    }

    private fun applyTemplate(input: List<String>): String {
        setEol("\n")
        val fw =
            FlywheelBuilder().inputList(input).withLineFeedRules(LINE_BREAKS).construct()
        return fw.exec()
    }
}
